bpo-34246: Use no mutable default args in smtplib

scorphus opened the pull request:
Some methods of the SMTP class use mutable default arguments. Specially `send_message` is affected in which it mutates one of the args by appending items to it, whic has side effects on further calls.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-34246](https://www.bugs.python.org/issue34246) -->
https://bugs.python.org/issue34246
<!-- /issue-number -->

scorphus committed:
bpo-34246: Use tuple as default arg for mail_options

Some methods of the SMTP class use mutable default arguments. Specially
`send_message` is affected in which it mutates one of the args by
appending items to it, whic has side effects on further calls.

serhiy-storchaka commented on Misc/NEWS.d/next/Library/2018-07-29-15-25-15.bpo-34246.xiKq-Q.rst:
```diff
@@ -0,0 +1,2 @@
+Use an empty tuple as the default argument for mail_options and avoid it from
```
The argument for what function? Using an empty tuple is an implementation detail, please describe the user visible behavior change. For example "```:meth:`smtplib.SMTP.send_message` no longer modifies the content of the *mail_options* argument.```" or something like.

Add also "Patch by Pablo S. Blum de Aguiar."

methane commented on Lib/smtplib.py:
```diff
@@ -958,7 +958,7 @@ def send_message(self, msg, from_addr=None, to_addrs=None,
                 g = email.generator.BytesGenerator(
                     bytesmsg, policy=msg.policy.clone(utf8=True))
-                mail_options += ['SMTPUTF8', 'BODY=8BITMIME']
+                mail_options = (*mail_options, 'SMTPUTF8', 'BODY=8BITMIME')
```
~I don't like this change.  `+=` seems more common/readable, and it's valid for immutable objects
like tuple and str.  (You need to change `[]` to `()` in this case)~

I'm sorry, I was wrong.  When caller passed list, list += tuple will raise TypeError.

scorphus committed:
Change NEWS file content as suggested

scorphus committed:
Add one more test case

Just to ensure `smtplib.SMTP.send_message` no longer modifies the
content of the *mail_options* argument

serhiy-storchaka reviewed (changes requested):
Please update also the signature in `Doc/library/smtplib.rst`.

scorphus committed:
Make rcpt_options compatible with what sendmail expects

scorphus committed:
Update method signature in Doc/library/smtplib.rst

scorphus committed:
Make empty tuple as values for options arguments

scorphus committed:
Use empty tuple as default value for `options` in `mail`

serhiy-storchaka reviewed (approved):
LGTM! Thank you for your contribution Pablo!

vstinner reviewed (approved):
LGTM.
